File: deploy/brev/visitor.js
const camera = document.querySelector('#camera');
const status = document.querySelector('#status');
const caption = document.querySelector('#caption');
let selected = 'worktop';
let selectedLabel = 'Worktop';
let generation = 0;
let timer;
let pending;
let displayedURL;
let displayedCamera;
let decodedAt = 0;
let frameHealthy = false;
let lastFrame;
let videoEnabled = false;
let nextVideoAttempt = 0;
const player = new window.VisitorVideo(document.querySelector('#video'), camera, () => {
  nextVideoAttempt = Date.now() + 5000;
  refresh();
});

caption.textContent = 'Connecting to Worktop camera…';

function cancelRefresh() {
  clearTimeout(timer);
  pending?.abort();
  pending = undefined;
}

async function refresh() {
  if (document.hidden || pending) return;
  if (player.playing) {
    timer = setTimeout(refresh, 1000);
    return;
  }
  const version = generation;
  const name = selected;
  const label = selectedLabel;
  const controller = new AbortController();
  pending = controller;
  const timeout = setTimeout(() => controller.abort(), 7000);
  let nextURL;
  let delay = 350;
  try {
    const response = await fetch(`/snapshot/${name}.jpg?t=${Date.now()}`, {
      cache: 'no-store', signal: controller.signal,
    });
    if (!response.ok) throw new Error('Camera unavailable');
    const blob = await response.blob();
    if (controller.signal.aborted || version !== generation) return;
    nextURL = URL.createObjectURL(blob);
    const frame = new Image();
    frame.src = nextURL;
    await frame.decode();
    if (controller.signal.aborted || version !== generation) return;

    const previousURL = displayedURL;
    camera.src = nextURL;
    displayedURL = nextURL;
    nextURL = undefined;
    displayedCamera = name;
    decodedAt = Date.now();
    frameHealthy = true;
    camera.alt = `${label} camera`;
    caption.textContent = `${label} · live camera`;
    if (previousURL) URL.revokeObjectURL(previousURL);
  } catch {
    if (version === generation) {
      frameHealthy = false;
      status.textContent = 'Camera reconnecting…';
    }
    delay = 2000;
  } finally {
    clearTimeout(timeout);
    if (nextURL) URL.revokeObjectURL(nextURL);
    if (pending === controller) {
      pending = undefined;
      timer = setTimeout(refresh, delay);
    }
  }
}

document.querySelectorAll('[data-camera]').forEach(button => {
  button.addEventListener('click', () => {
    generation += 1;
    cancelRefresh();
    player.stop();
    selected = button.dataset.camera;
    selectedLabel = button.textContent.trim();
    lastFrame = undefined;
    frameHealthy = false;
    status.textContent = `Loading ${selectedLabel} camera…`;
    document.querySelectorAll('[data-camera]').forEach(item => item.setAttribute('aria-pressed', String(item === button)));
    refresh();
    if (videoEnabled) player.start(selected);
  });
});
document.addEventListener('visibilitychange', refresh);
window.addEventListener('pagehide', () => {
  generation += 1;
  cancelRefresh();
  player.stop();
  if (displayedURL) URL.revokeObjectURL(displayedURL);
  displayedURL = undefined;
});
window.addEventListener('pageshow', refresh);

async function updateStatus() {
  const version = generation;
  try {
    const response = await fetch('/api/status', {cache: 'no-store', signal: AbortSignal.timeout(7000)});
    if (!response.ok) throw new Error('Camera unavailable');
    const state = await response.json();
    if (version !== generation) return;
    videoEnabled = state.video?.enabled === true;
    if (videoEnabled && !player.camera && Date.now() >= nextVideoAttempt) player.start(selected);
    const row = state.cameras.find(item => item.name === selected);
    const ready = player.playing || (frameHealthy && displayedCamera === selected && Date.now() - decodedAt < 7000);
    const advancing = row?.online && row.frame_id != null && row.frame_id !== lastFrame;
    status.textContent = ready && advancing ? (player.playing ? 'Live · H.264 video' : `Live · frame ${row.frame_id}`)
      : displayedCamera !== selected ? `Loading ${selectedLabel} camera…`
      : !frameHealthy ? 'Camera reconnecting…' : 'Waiting for a fresh frame…';
    lastFrame = row?.frame_id;
  } catch {
    if (version === generation) status.textContent = 'Camera reconnecting…';
  } finally {
    setTimeout(updateStatus, 3000);
  }
}
refresh();
updateStatus();

window.addEventListener('offline', () => {
  player.stop();
  nextVideoAttempt = Infinity;
  status.textContent = 'Camera reconnecting…';
});
window.addEventListener('online', () => {
  nextVideoAttempt = 0;
  refresh();
});

File: deploy/brev/visitor.py
"""Serve a small public camera view without forwarding administrative routes."""

from __future__ import annotations

import argparse
import base64
import hmac
from itertools import chain
import json
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
import threading
from urllib.error import URLError
from urllib.parse import urlsplit
from urllib.request import urlopen

CAMERAS = ("kitchen", "worktop", "side")
ROOT = Path(__file__).resolve().parent


class VisitorServer(ThreadingHTTPServer):
    daemon_threads = True
    request_queue_size = 16

    def __init__(self, address, handler, max_connections=32):
        self.slots = threading.BoundedSemaphore(max_connections)
        super().__init__(address, handler)

    def get_request(self):
        connection, address = super().get_request()
        connection.settimeout(10)
        return connection, address

    def process_request(self, request, address):
        if not self.slots.acquire(blocking=False):
            try:
                request.sendall(b"HTTP/1.1 503 Service Unavailable\r\nContent-Length: 0\r\nConnection: close\r\nRetry-After: 2\r\n\r\n")
            except OSError:
                pass
            self.shutdown_request(request)
            return
        try:
            super().process_request(request, address)
        except BaseException:
            self.slots.release()
            raise

    def process_request_thread(self, request, address):
        try:
            super().process_request_thread(request, address)
        finally:
            self.slots.release()


class VisitorHandler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    def log_message(self, format, *args):
        pass

    def respond(self, status, body, content_type="application/json"):
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("Referrer-Policy", "no-referrer")
        self.send_header("Content-Security-Policy", "default-src 'self'; img-src 'self' blob:; media-src 'self' blob:; style-src 'self'; script-src 'self'; frame-ancestors 'none'")
        if status == 401:
            self.send_header("WWW-Authenticate", 'Basic realm="Physical Agentic AI", charset="UTF-8"')
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self):
        if self.server.authorization and not hmac.compare_digest(
            self.headers.get("Authorization", ""), self.server.authorization
        ):
            return self.respond(401, b'{"error":"Authentication required"}')
        path = urlsplit(self.path).path
        if path in ("/", "/visitor.css", "/visitor.js", "/visitor-player.js",
                    "/staff/", "/staff/style.css", "/staff/media/openclaw-cameras.png"):
            name, content_type = {
                "/": ("visitor.html", "text/html; charset=utf-8"),
                "/visitor.css": ("visitor.css", "text/css; charset=utf-8"),
                "/visitor.js": ("visitor.js", "text/javascript; charset=utf-8"),
                "/visitor-player.js": ("visitor-player.js", "text/javascript; charset=utf-8"),
                "/staff/": ("staff.html", "text/html; charset=utf-8"),
                "/staff/style.css": ("staff.css", "text/css; charset=utf-8"),
                "/staff/media/openclaw-cameras.png": ("staff-openclaw-cameras.png", "image/png"),
            }[path]
            return self.respond(200, (ROOT / name).read_bytes(), content_type)
        try:
            if path in {f"/video/{name}.mp4" for name in CAMERAS}:
                return self.video(path.split("/")[-1][:-4])
            if path == "/api/status":
                with urlopen(self.server.camera_origin + "/state", timeout=5) as response:
                    state = json.load(response)
                rows = state.get("cameras", [])
                rows = {row["name"]: row for row in rows} if isinstance(rows, list) else rows
                cameras = []
                for name in CAMERAS:
                    row = rows.get(name, {})
                    cameras.append({
                        "name": name, "label": name.capitalize(),
                        "online": row.get("online", False),
                        "frame_id": row.get("frame_id"),
                        "fps": row.get("fps"),
                        "stream_url": f"/snapshot/{name}.jpg",
                    })
                return self.respond(200, json.dumps({"cameras": cameras,
                    "video": {"enabled": getattr(self.server, "video", None) is not None,
                              "codec": "h264", "nominal_fps": 3},
                    "camera_discovery": {"version": 1, "base_url": "/", "state_url": "/api/status",
                        "default_camera": "worktop", "cameras": cameras}}).encode())
            if path in {f"/snapshot/{name}.jpg" for name in CAMERAS}:
                with urlopen(self.server.camera_origin + path, timeout=5) as response:
                    frame = response.read(4 * 1024 * 1024)
                if not frame.startswith(b"\xff\xd8") or not frame.endswith(b"\xff\xd9"):
                    raise ValueError("Camera did not return JPEG")
                return self.respond(200, frame, "image/jpeg")
        except (URLError, OSError, ValueError):
            return self.respond(503, b'{"error":"Camera is reconnecting"}')
        return self.respond(404, b'{"error":"Not found"}')

    def video(self, camera):
        streams = getattr(self.server, "video", None)
        if streams is None:
            return self.respond(404, b'{"error":"Video is disabled"}')
        try:
            chunks = streams.get(camera).chunks()
            initial = next(chunks)
        except (OSError, ValueError, TimeoutError):
            return self.respond(503, b'{"error":"Video is reconnecting"}')
        self.send_response(200)
        self.send_header("Content-Type", "video/mp4")
        self.send_header("Transfer-Encoding", "chunked")
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("X-Accel-Buffering", "no")
        self.end_headers()
        self.connection.settimeout(10)
        try:
            for chunk in chain((initial,), chunks):
                self.wfile.write(f"{len(chunk):x}\r\n".encode() + chunk + b"\r\n")
                self.wfile.flush()
        except (OSError, TimeoutError):
            pass
        finally:
            self.close_connection = True

    def do_POST(self):
        self.respond(405, b'{"error":"Method not allowed"}')


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--port", type=int, default=8093)
    parser.add_argument("--camera-origin", default="http://127.0.0.1:8091")
    parser.add_argument("--auth-file", type=Path)
    parser.add_argument("--video", action="store_true", help="Share the camera feeds as NVENC H.264 video")
    parser.add_argument("--ffmpeg", default="ffmpeg")
    args = parser.parse_args()
    server = VisitorServer(("127.0.0.1", args.port), VisitorHandler)
    server.camera_origin = args.camera_origin.rstrip("/")
    server.video = None
    if args.video:
        from visitor_video import VideoStreams
        server.video = VideoStreams(server.camera_origin, args.ffmpeg)
    server.authorization = ""
    if args.auth_file:
        credentials = json.loads(args.auth_file.read_text())
        value = f"{credentials['username']}:{credentials['password']}".encode()
        server.authorization = "Basic " + base64.b64encode(value).decode()
    try:
        server.serve_forever()
    finally:
        if server.video:
            server.video.stop()
        server.server_close()


if __name__ == "__main__":
    main()
